### ISSUE TITLE
[Dev Containers] Settings to control ripgrep exclusions for devcontainer.json

### ISSUE THREAD

**Issue by sam-mfb** [microsoft/vscode-remote-release#9738]
There are a number of issues filed against the Dev Container extension relating to ripgrep (i.e., `rg` or `rg.exe`) causing slow performance, high memory usage, or battery drains (on laptops) when it scan for `devcontainer.json` files. 

Some of the more extensively commented open issues on this include: [8661](https://github.com/microsoft/vscode-remote-release/issues/8661) and [9110](https://github.com/microsoft/vscode-remote-release/issues/9110). However, the issue has also been raised in the following issues (some of which are closed, but only because of inactivity/lack of info, not because they were resolved) [9586](https://github.com/microsoft/vscode-remote-release/issues/9586), [9699](https://github.com/microsoft/vscode-remote-release/issues/9586), [8635](https://github.com/microsoft/vscode-remote-release/issues/8635)

All of these problems in general appear to stem from the fact that the `rg` command used to scan for `devcontainer.json` files can get bogged down scanning large repos or repos with extensive symlinks, even when it is unlikely that those locations will actually contain `devcontainer.json` files.

The `rg` command that the Dev Container extension runs does seem to try to exclude certain locations that are unlikely to contain `devcontainer.json` files. The command appears to be (according to `ps`):
```
rg --files --hidden --case-sensitive --no-require-git -g /{.devcontainer.json,.devcontainer/devcontainer.json,.devcontainer/*/devcontainer.json} -g !**/.git -g !**/.svn -g !**/.hg -g !**/CVS -g !**/.DS_Store -g !**/Thumbs.db --no-ignore --follow --no-config --no-ignore-global
```
However, these exclusions are not enough to prevent the slow down behavior. For example, in a node repository (particularly, a monorepo) the `node_modules` folder can get quite big once it is populated and this `rg` call will bog down scanning it all even though it is unlikely to contain any `devcontainer.json` files. Some of the tickets above contain other examples, but I believe in general the behavior will occur in any decently sized repo (even if the size comes from things that are not committed to version control, like `node_modules`)

Various solutions have been proposed in some of the tickets above, e.g., restricting `rg` to a `--max-depth` or ignoring files in `.gitignore`. The problem, I assume, is that there could be legitimate reasons for those to be included -- a `devcontainer.json` file could be legitimately included at an arbitrary depth now that multiple devcontainers per repo are supported; a repo may want to exclude devcontainers from version control if developers are encouraged to create their own containers.

One of the tickets also suggests using `files.exclude,` but it is not clear why the above `rg` command would respect that, and, in any event, that may not be the desired behavior, i.e., the files may need to be searchable for other reasons, just not to locate a devcontainer.

It seems to me the best solution would give the end-user direct control over what gets excluded when searching for `devcontainer.json` files. This could be achieved by a user setting in the Dev Container extension called something like "Exclude paths from devcontainer search:", which would accept an array of paths. Alternatively a dotfile could be used, e.g., `.devcontainerignore`, that contained a list of paths to exclude. The latter would be nice as it could be committed to version control.

**Comment by chrmarti** [comment]
Thanks for the detailed report. Apparently `/{.devcontainer.json,.devcontainer/devcontainer.json,.devcontainer/*/devcontainer.json}` doesn't prune the folder traversal of `rg`. I'll investigate if the search can be limited to the `.devcontainer` folder by using a `RelativePattern`. /cc @andreamah

**Comment by chrmarti** [comment]
@sam-mfb This fix is available in Dev Containers 0.356.0-pre-release. Could you give that a try and let me know if that fixes it? Thanks!

**Comment by sam-mfb** [comment]
@chrmarti, tested and this works. with node_modules downloaded, our monorepo would hang for several minutes with rg using max cpu when you tried to "Reopen in Container". With the pre-release version it opens instantly, no rg spike. Thank you!
